On CUDA **sm_53** `non-constant-expression cannot be narrowed from type 'unsigned long' to 'uint32_t' (aka 'unsigned int') in initializer list`

**Bug summary**

Hello Aksel, today I was targeting CUDA platform `sm_53` using `syclcc`, with following trivial code

```cpp
#include <CL/sycl.hpp>
#include <iostream>

int main(int argc, char **argv) {
  cl::sycl::device d{cl::sycl::default_selector{}};
  cl::sycl::queue q{d};

  std::cout << "running on " << d.get_info<cl::sycl::info::device::name>()
            << std::endl;

  return 0;
}
```

and I get

```bash
syclcc-clang -std=c++17 -O3 -c main.cpp

clang: warning: Unknown CUDA version 10.2. Assuming the latest supported version 10.1 [-Wunknown-cuda-version]
In file included from main.cpp:1:
In file included from /home/anjan/.local/hipsycl/bin/../include/CL/sycl.hpp:31:
In file included from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:88:
In file included from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/queue.hpp:47:
In file included from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/handler.hpp:65:
In file included from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/glue/kernel_launcher_factory.hpp:43:
In file included from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/glue/cuda/cuda_kernel_launcher.hpp:31:
In file included from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/glue/cuda/../generic/hiplike/hiplike_kernel_launcher.hpp:53:
In file included from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/sycl/interop_handle.hpp:34:
In file included from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/backend_interop.hpp:37:
/home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/kernel.hpp:107:23: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'uint32_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]
      return uint32_t{dev.get_info<info::device::max_work_group_size>() /
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/kernel.hpp:107:23: note: insert an explicit cast to silence this issue
      return uint32_t{dev.get_info<info::device::max_work_group_size>() /
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                      static_cast<uint32_t>(
/home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/kernel.hpp:118:23: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'uint32_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]
      return uint32_t{*std::max_element(subgroups.begin(), subgroups.end())};
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/kernel.hpp:118:23: note: insert an explicit cast to silence this issue
      return uint32_t{*std::max_element(subgroups.begin(), subgroups.end())};
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                      static_cast<uint32_t>(                               )
2 errors generated when compiling for sm_53.
make: *** [Makefile:9: main.o] Error 1
```

**To Reproduce**

Take aforementioned cpp snippet, put it in `main.cpp` while also copy following in Makefile

```makefile
CXX = syclcc
CXXFLAGS = -std=c++17 -O3
PROG = run

$(PROG): main.o
	$(CXX) $(HIPSYCLFLAGS) $^ -o $@

main.o: main.cpp
	$(CXX) $(CXXFLAGS) -c $^
```

Then run

```bash
make
```

**Expected behavior**

Compilation must not fail anyway.

**Describe your setup**

Compiled from source at [commit](https://github.com/illuhad/hipSYCL/commit/cc794e3291b5f365d6241689da16b5098f4a7a4a)

* Describe the dependencies that hipSYCL sits on top of in your setup: 
   * clang version and how clang was installed `clang version 10.0.0-4ubuntu1`
   * host compiler `c++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0`
   * CUDA version 10.2 ( though compiler falls back to 10.1 )
   * I'm on aarch64 machine, to be more specific it's Nvidia Jetson Nano

**Additional context**

The issue is in [here](https://github.com/illuhad/hipSYCL/blob/cc794e3291b5f365d6241689da16b5098f4a7a4a/include/hipSYCL/sycl/kernel.hpp#L107-L108) & [here](https://github.com/illuhad/hipSYCL/blob/cc794e3291b5f365d6241689da16b5098f4a7a4a/include/hipSYCL/sycl/kernel.hpp#L118), where implicit type conversion happening.

One thing I found, when I explicitly ask `syclcc` to compile targeting `omp`, I get following warnings 

```bash
syclcc -std=c++17 -O3 --hipsycl-targets='omp' -c main.cpp
In file included from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/sycl/backend_interop.hpp:37,
                 from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/sycl/interop_handle.hpp:34,
                 from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/glue/omp/omp_kernel_launcher.hpp:43,
                 from /home/anjan/.local/hipsycl/bin/../include/hipSYCL/glue/kernel_launcher_factory.hpp:47,
                 from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/handler.hpp:65,
                 from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/queue.hpp:47,
                 from /home/anjan/.local/hipsycl/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:88,
                 from /home/anjan/.local/hipsycl/bin/../include/CL/sycl.hpp:31,
                 from main.cpp:1:
/home/anjan/.local/hipsycl/bin/../include/hipSYCL/sycl/kernel.hpp: In member function ‘typename KernelDeviceSpecificT::return_type hipsycl::sycl::kernel::get_info(const hipsycl::sycl::device&) const’:
/home/anjan/.local/hipsycl/bin/../include/hipSYCL/sycl/kernel.hpp:107:73: warning: narrowing conversion of ‘(dev.get_info<hipsycl::sycl::info::device::max_work_group_size>() / std::min_element(subgroups.std::vector<long unsigned int>::begin(), subgroups.std::vector<long unsigned int>::end()).__gnu_cxx::__normal_iterator<long unsigned int*, std::vector<long unsigned int> >::operator*())’ from ‘hipsycl::sycl::info::param_traits<hipsycl::sycl::info::device, hipsycl::sycl::info::device::max_work_group_size>::return_type’ {aka ‘long unsigned int’} to ‘uint32_t’ {aka ‘unsigned int’} [-Wnarrowing]
  107 |       return uint32_t{dev.get_info<info::device::max_work_group_size>() /
      |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  108 |                       (*std::min_element(subgroups.begin(), subgroups.end()))};
      |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/anjan/.local/hipsycl/bin/../include/hipSYCL/sycl/kernel.hpp:118:76: warning: narrowing conversion of ‘std::max_element<__gnu_cxx::__normal_iterator<long unsigned int*, std::vector<long unsigned int> > >(subgroups.std::vector<long unsigned int>::begin(), subgroups.std::vector<long unsigned int>::end()).__gnu_cxx::__normal_iterator<long unsigned int*, std::vector<long unsigned int> >::operator*()’ from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} [-Wnarrowing]
  118 |       return uint32_t{*std::max_element(subgroups.begin(), subgroups.end())};
      |                                                                            ^
syclcc --hipsycl-targets='omp' main.o -o run
```

But finally it produces executable binary, which can be run fine

```bash
./run

running on hipSYCL OpenMP host device
```

--- a/include/hipSYCL/sycl/kernel.hpp
+++ b/include/hipSYCL/sycl/kernel.hpp
@@ -104,8 +104,9 @@ public:
                                         max_num_sub_groups>) {
                                           
       auto subgroups = dev.get_info<info::device::sub_group_sizes>();
-      return uint32_t{dev.get_info<info::device::max_work_group_size>() /
-                      (*std::min_element(subgroups.begin(), subgroups.end()))};
+      return static_cast<uint32_t>(
+          dev.get_info<info::device::max_work_group_size>() /
+          (*std::min_element(subgroups.begin(), subgroups.end())));
 
     } else if constexpr (std::is_same_v<KernelDeviceSpecificT,
                                         compile_num_sub_groups>) {
@@ -115,8 +116,9 @@ public:
                                         max_sub_group_size>) {
 
       auto subgroups = dev.get_info<info::device::sub_group_sizes>();
-      return uint32_t{*std::max_element(subgroups.begin(), subgroups.end())};
-      
+      return static_cast<uint32_t>(
+          *std::max_element(subgroups.begin(), subgroups.end()));
+
     } else if constexpr (std::is_same_v<KernelDeviceSpecificT,
                                         compile_sub_group_size>) {
       return uint32_t{0};

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -65,6 +65,7 @@ add_executable(sycl_tests
   sycl/group_functions/group_functions_reduce.cpp
   sycl/group_functions/group_functions_scan.cpp
   sycl/id_range.cpp
+  sycl/info_queries.cpp
   sycl/item.cpp
   sycl/kernel_invocation.cpp
   sycl/math.cpp

--- a/tests/sycl/info_queries.cpp
+++ b/tests/sycl/info_queries.cpp
@@ -1,0 +1,68 @@
+/*
+ * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ *
+ * Copyright (c) 2021 Aksel Alpay
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "sycl_test_suite.hpp"
+using namespace cl;
+
+BOOST_FIXTURE_TEST_SUITE(info_queries, reset_device_fixture)
+
+BOOST_AUTO_TEST_CASE(device_queries) {
+  
+  sycl::device d{sycl::default_selector{}};
+  sycl::queue q{d};
+
+  std::string device = d.get_info<sycl::info::device::name>();
+  BOOST_TEST(device.length() > 0);
+  // TODO Add tests for more queries
+}
+BOOST_AUTO_TEST_CASE(kernel_specific_queries) {
+  
+  sycl::queue q;
+
+  sycl::program p{q.get_context()};
+  sycl::kernel k = p.get_kernel<void>();
+
+  auto wg_size =
+      k.get_info<sycl::info::kernel_device_specific::work_group_size>(
+          q.get_device());
+  BOOST_TEST(wg_size > 0);
+
+  auto max_sgs =
+      k.get_info<sycl::info::kernel_device_specific::max_num_sub_groups>(
+          q.get_device());
+  BOOST_TEST(max_sgs > 0);
+
+  auto max_sg_size =
+      k.get_info<sycl::info::kernel_device_specific::max_sub_group_size>(
+          q.get_device());
+  BOOST_TEST(max_sg_size > 0);
+  
+  // TODO Add tests for more queries
+}
+
+BOOST_AUTO_TEST_SUITE_END()

